schema/function_test.go

```diff
@@ -0,0 +1,517 @@
+package schema_test
+
+import (
+	"fmt"
+	"go.arcalot.io/assert"
+	"go.flow.arcalot.io/pluginsdk/schema"
+	"reflect"
+	"testing"
+)
+
+// Simple, no input or output.
+func TestCallableFunctionSchema_Simple(t *testing.T) {
+	called := false
+	simpleFunc, err := schema.NewCallableFunction(
+		"test",
+		make([]schema.Type, 0),
+		nil,
+		nil,
+		func() {
+			called = true
+		},
+	)
+	assert.NoError(t, err)
+	assert.Equals(t, called, false)
+	result, err := simpleFunc.Call([]any{})
+	assert.NoError(t, err)
+	assert.Nil(t, result)
+	assert.Equals(t, called, true)
+}
+
+// Simple, with one input.
+func TestCallableFunctionSchema_Simple1Param(t *testing.T) {
+	passedInVal := ""
+	simpleFunc, err := schema.NewCallableFunction(
+		"test",
+		[]schema.Type{schema.NewStringSchema(nil, nil, nil)},
+		nil,
+		nil,
+		func(test string) {
+			passedInVal = test
+		},
+	)
+	assert.NoError(t, err)
+	result, err := simpleFunc.Call([]any{"a"})
+	assert.NoError(t, err)
+	assert.Nil(t, result)
+	assert.Equals(t, passedInVal, "a")
+}
+
+// Simple, with no input, and only error out.
+func TestCallableFunctionSchema_SimpleWithNilErr(t *testing.T) {
+	simpleFunc, err := schema.NewCallableFunction(
+		"test",
+		make([]schema.Type, 0),
+		nil,
+		nil,
+		func() error {
+			return nil
+		},
+	)
+	assert.NoError(t, err)
+	result, err := simpleFunc.Call([]any{})
+	// No result or error
+	assert.NoError(t, err)
+	assert.Nil(t, result)
+}
+
+// One input, nil error output.
+func TestCallableFunctionSchema_1ParamWithNilErr(t *testing.T) {
+	passedInVal := ""
+	simpleFunc, err := schema.NewCallableFunction(
+		"test",
+		[]schema.Type{schema.NewStringSchema(nil, nil, nil)},
+		nil,
+		nil,
+		func(test string) error {
+			passedInVal = test
+			return nil
+		},
+	)
+	assert.NoError(t, err)
+	result, err := simpleFunc.Call([]any{"a"})
+	assert.NoError(t, err)
+	assert.Nil(t, result)
+	// Validate that the schema.Type got passed in properly.
+	assert.Equals(t, passedInVal, "a")
+}
+
+// Simple, with error out.
+func TestCallableFunctionSchema_SimpleWithErr(t *testing.T) {
+	simpleFunc, err := schema.NewCallableFunction(
+		"test",
+		make([]schema.Type, 0),
+		nil,
+		nil,
+		func() error {
+			return fmt.Errorf("this is an error")
+		},
+	)
+	assert.NoError(t, err)
+	_, err = simpleFunc.Call([]any{})
+	// There should be an error from the function. Validate that it's the correct one.
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "this is an error")
+}
+
+// Simple, one output.
+func TestCallableFunctionSchema_OneReturn(t *testing.T) {
+	simpleFunc, err := schema.NewCallableFunction(
+		"test",
+		make([]schema.Type, 0),
+		schema.NewStringSchema(nil, nil, nil),
+		nil,
+		func() string {
+			return "a"
+		},
+	)
+	assert.NoError(t, err)
+	result, err := simpleFunc.Call([]any{})
```

Comment:
This is generic, but also it could return an error if there is an incorrect number of args passed in, or if the value stored in the struct was not validated (by passing the checks in `NewCallableFunction` or `NewDynamicCallableFunction`).

`TestCallableFunctionSchema_Err_CallWrongErrorReturn` and `TestCallableFunctionSchema_Err_CallWrongReturnCount` test bypassing the checks.